### PR TITLE
Refactor Error Handling

### PR DESCRIPTION
**Pull request by PineappleIOnic**
This PR refactors how Errors are handled to allow for easier error reporting from sources.

This PR must be released as a new minor version to maintain compatibility with previous Appwrite versions.

It starts to move away from the `cache` being the single source of truth surrounding a migration by allowing both Source and Destination to store an array of errors both encountered. This fixes an issue that made it hard to keep a detailed record of issues that sources ran into as errors are now no longer tied to a specific resource

### CODE REVIEW

File: src/Migration/Error.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Utopia\Migration;
+
+class Error extends \Exception
```
Comment (christyjacob4) [inline src/Migration/Error.php:5]: Why error? Why not call it `Exception` itself?

File: src/Migration/Resource.php [hunk 2]
```diff
@@ -52,7 +52,7 @@ abstract class Resource
 
     public const TYPE_HASH = 'hash';
 
-    public const TYPE_ENVVAR = 'envvar';
+    public const TYPE_ENVVAR = 'environment variable';
```
Comment (christyjacob4) [inline src/Migration/Resource.php:55]: As eldad mentioned, in another PR we need to refactor this constant and give it the full type name `TYPE_ENVIRONMENT_VARIABLE`

File: src/Migration/Sources/Appwrite.php [hunk 3]
```diff
@@ -977,15 +978,24 @@ private function exportFiles(int $batchSize)
...
+                        $this->exportFileData(new File(
+                            $file['$id'],
+                            $bucket,
+                            $file['name'],
+                            $file['signature'],
+                            $file['mimeType'],
+                            $file['$permissions'],
+                            $file['sizeOriginal'],
+                        ));
+                    } catch (\Exception $e) {
```
Comment (christyjacob4) [inline src/Migration/Sources/Appwrite.php:991]:
```suggestion
                    } catch (\Throwable $th) {
```

Make sure to catch Throwable everywhere. And then create your own exception out of it

File: src/Migration/Resource.php (hunk 2 above)
Comment (PineappleIOnic) [reply src/Migration/Resource.php:55]: In another PR? I'll create a task

File: src/Migration/Sources/Appwrite.php [hunk 4]
```diff
@@ -977,15 +979,24 @@ private function exportFiles(int $batchSize)
...
+                            $file['$id'],
+                            $bucket,
+                            $file['name'],
+                            $file['signature'],
+                            $file['mimeType'],
+                            $file['$permissions'],
+                            $file['sizeOriginal'],
+                        ));
+                    } catch (\Throwable $e) {
+                        $this->pushError(new Exception(
```
Comment (eldadfux) [inline src/Migration/Sources/Appwrite.php:993]: Name is correct, but we would usually use a name like `addError` in utopia

Comment (PineappleIOnic) [reply src/Migration/Sources/Appwrite.php:993]: Alright, will alter accordingly

**Review by eldadfux — APPROVED** [review]

**Comment by PineappleIOnic** [comment]
Need to get #26 merged into `dev` for tests to pass here

**Review by Meldiron — APPROVED** [review]

**Pull request merged**